pkg/build/strategies/dockerfile/dockerfile.go

```diff
@@ -0,0 +1,327 @@
+			buffer.WriteString(fmt.Sprintf("%s=%s", k, v))
+			first = false
+		}
+		buffer.WriteString("\n")
+	}
+
+	// only COPY scripts dir if required scripts are present, i.e. the dir is not empty;
+	// even if the "scripts" dir exists, the COPY would fail if it was empty
+	scriptsProvided, fileNames := checkValidDirWithContents(filepath.Join(config.WorkingDir, builder.uploadScriptsDir))
+	assembleProvided := false
+	runProvided := false
+	for _, f := range fileNames {
+		glog.V(2).Infof("found override script file %s", f.Name())
+		if f.Name() == "run" {
+			runProvided = true
+		} else if f.Name() == "assemble" {
+			assembleProvided = true
+		}
+		if runProvided && assembleProvided {
+			break
+		}
+	}
+
+	env := sti.CreateBuildEnvironment(config.WorkingDir, config.Environment)
+	if len(env) > 0 {
+		first := true
+		buffer.WriteString("ENV ")
+		for _, v := range env {
+			if !first {
+				buffer.WriteString(fmt.Sprintf(" \\\n    "))
+			}
+			buffer.WriteString(fmt.Sprintf("%s", sanitize(v)))
+			first = false
+		}
+		buffer.WriteString("\n")
+	}
+
+	if scriptsProvided {
+		glog.V(2).Infof("Override scripts are included in directory %q", builder.uploadScriptsDir)
+		buffer.WriteString(fmt.Sprintf("COPY --chown=%s:0 %s %s\n", sanitize(imageUser), sanitize(builder.uploadScriptsDir), sanitize(filepath.ToSlash(scriptsDestDir))))
+	} else {
+		// if an err on reading or opening dir, can't copy it
+		glog.V(2).Infof("Could not gather scripts from the directory %q", builder.uploadScriptsDir)
+	}
+
+	// copy in the user's source code.
+	buffer.WriteString(fmt.Sprintf("COPY --chown=%s:0 %s %s\n", sanitize(imageUser), sanitize(builder.uploadSrcDir), sanitize(filepath.ToSlash(sourcesDestDir))))
+
+	glog.V(2).Infof("Processing secret inputs: %#v", config.Injections)
+	config.Injections = util.FixInjectionsWithRelativePath(config.ImageWorkDir, config.Injections)
+	glog.V(2).Infof("Processed secret inputs: %#v", config.Injections)
+
+	for _, injection := range config.Injections {
+		src := filepath.Join(api.Injections, injection.Source)
+		buffer.WriteString(fmt.Sprintf("COPY --chown=%s:0 %s %s\n", sanitize(imageUser), sanitize(src), sanitize(injection.Destination)))
+	}
+
+	if assembleProvided {
+		buffer.WriteString(fmt.Sprintf("RUN %s\n", sanitize(filepath.Join(scriptsDestDir, "assemble"))))
+	} else {
+		buffer.WriteString(fmt.Sprintf("# Assemble script sourced from builder image based on user input or image metadata.\n"))
+		buffer.WriteString(fmt.Sprintf("# If this file does not exist in the image, the build will fail.\n"))
+		buffer.WriteString(fmt.Sprintf("RUN %s\n", sanitize(filepath.Join(config.ImageScriptsDir, "assemble"))))
+	}
+
+	// Remove any secrets that were copied into the image,
+	// but leave configmap content alone.
+	wroteRun := false
+	for _, injection := range config.Injections {
+		if injection.Keep {
+			continue
+		}
+		if !wroteRun {
+			buffer.WriteString("RUN ")
+			wroteRun = true
+		} else {
+			buffer.WriteString(" && ")
```

Comment:
/s/" && "/" \\\n&& "/
interesting note - Docker's ["best practice" example](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run) doesn't indent the `&&` with multiline

Reply:
new commit pushed, here's the latest output:

```
FROM docker.io/centos/nodejs-8-centos7
USER 1001
LABEL io.openshift.s2i.build.image=docker.io/centos/nodejs-8-centos7 \
      io.openshift.s2i.build.commit.author=Honza Horak <hhorak@redhat.com> \
      io.openshift.s2i.build.commit.date=Fri May 4 09:16:28 2018 +0200 \
      io.openshift.s2i.build.commit.id=a2aa814098856b1d935f68db1eb808d091e1a44e \
      io.openshift.s2i.build.commit.ref=master \
      io.openshift.s2i.build.commit.message=Merge pull request #178 from wozniakjan/issue/origin_19379/generalization_of_p.. \
      io.openshift.s2i.build.source-location=https://github.com/openshift/nodejs-ex
ENV FOO=BAR \
    BAR=FOO
COPY --chown=1001:0 upload/src /tmp/src
COPY --chown=1001:0 upload/injections/tmp/s /tmp/s
COPY --chown=1001:0 upload/injections/tmp/s /tmp/s2
# Assemble script sourced from builder image based on user input or image metadata.
# If this file does not exist in the image, the build will fail.
RUN /usr/libexec/s2i/assemble
RUN rm -rf /tmp/s && \
    rm -rf /tmp/s2
# Run script sourced from builder image based on user input or image metadata.
# If this file does not exist in the image, the build will fail.
CMD /usr/libexec/s2i/run
```